### PR TITLE
Allow easy styling autofilled of Inputs

### PR DESCRIPTION
**Pull request by hasparus**
[Image from Gyazo](https://gyazo.com/d7f0e0f412043b96a8f775c618255fe4)

## Release Notes

Background color of autofilled inputs will now default to `theme.colors.background`. It can be changed by setting `autofillBackgroundColor` prop.

<details>
  <summary>📦 Published PR as canary version: <code>0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0</code></summary>
  <br />
  
  :sparkles: Test out this PR locally via:
  
  ```bash
  npm install @theme-ui/color-modes@0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0
  npm install @theme-ui/color@0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0
  npm install @theme-ui/components@0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0
  npm install @theme-ui/core@0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0
  npm install @theme-ui/css@0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0
  npm install @theme-ui/custom-properties@0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0
  npm install @theme-ui/editor@0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0
  npm install gatsby-plugin-theme-ui@0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0
  npm install gatsby-theme-style-guide@0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0
  npm install gatsby-theme-ui-layout@0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0
  npm install @theme-ui/match-media@0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0
  npm install @theme-ui/mdx@0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0
  npm install @theme-ui/parse-props@0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0
  npm install @theme-ui/preset-base@0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0
  npm install @theme-ui/preset-bootstrap@0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0
  npm install @theme-ui/preset-bulma@0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0
  npm install @theme-ui/preset-dark@0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0
  npm install @theme-ui/preset-deep@0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0
  npm install @theme-ui/preset-funk@0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0
  npm install @theme-ui/preset-future@0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0
  npm install @theme-ui/preset-polaris@0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0
  npm install @theme-ui/preset-roboto@0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0
  npm install @theme-ui/preset-sketchy@0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0
  npm install @theme-ui/preset-swiss@0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0
  npm install @theme-ui/preset-system@0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0
  npm install @theme-ui/preset-tailwind@0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0
  npm install @theme-ui/preset-tosh@0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0
  npm install @theme-ui/presets@0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0
  npm install @theme-ui/prism@0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0
  npm install @theme-ui/sidenav@0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0
  npm install @theme-ui/style-guide@0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0
  npm install @theme-ui/tachyons@0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0
  npm install @theme-ui/tailwind@0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0
  npm install @theme-ui/theme-provider@0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0
  npm install theme-ui@0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0
  npm install @theme-ui/typography@0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0
  # or 
  yarn add @theme-ui/color-modes@0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0
  yarn add @theme-ui/color@0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0
  yarn add @theme-ui/components@0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0
  yarn add @theme-ui/core@0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0
  yarn add @theme-ui/css@0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0
  yarn add @theme-ui/custom-properties@0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0
  yarn add @theme-ui/editor@0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0
  yarn add gatsby-plugin-theme-ui@0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0
  yarn add gatsby-theme-style-guide@0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0
  yarn add gatsby-theme-ui-layout@0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0
  yarn add @theme-ui/match-media@0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0
  yarn add @theme-ui/mdx@0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0
  yarn add @theme-ui/parse-props@0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0
  yarn add @theme-ui/preset-base@0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0
  yarn add @theme-ui/preset-bootstrap@0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0
  yarn add @theme-ui/preset-bulma@0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0
  yarn add @theme-ui/preset-dark@0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0
  yarn add @theme-ui/preset-deep@0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0
  yarn add @theme-ui/preset-funk@0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0
  yarn add @theme-ui/preset-future@0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0
  yarn add @theme-ui/preset-polaris@0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0
  yarn add @theme-ui/preset-roboto@0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0
  yarn add @theme-ui/preset-sketchy@0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0
  yarn add @theme-ui/preset-swiss@0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0
  yarn add @theme-ui/preset-system@0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0
  yarn add @theme-ui/preset-tailwind@0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0
  yarn add @theme-ui/preset-tosh@0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0
  yarn add @theme-ui/presets@0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0
  yarn add @theme-ui/prism@0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0
  yarn add @theme-ui/sidenav@0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0
  yarn add @theme-ui/style-guide@0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0
  yarn add @theme-ui/tachyons@0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0
  yarn add @theme-ui/tailwind@0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0
  yarn add @theme-ui/theme-provider@0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0
  yarn add theme-ui@0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0
  yarn add @theme-ui/typography@0.10.0-canary.1811.dd54586b88593673789513c59bcbe0235916cbf1.0
  ```
</details>

### CODE REVIEW

File: packages/components/src/Input.js [hunk 1]
```diff
@@ -1,26 +1,52 @@
...
+  color: 'inherit',
+  bg: 'transparent',
+
+  ':autofill, :autofill:hover, :autofill:focus': autofillStyles,
+  ':-webkit-autofill, :-webkit-autofill:hover, :-webkit-autofill:focus':
+    autofillStyles,
+}
+
+export const Input = React.forwardRef(function Input(
+  { sx, autofillBackgroundColor = 'background', ...rest },
```
Comment (lachlanjc) [inline packages/components/src/Input.js:34]: Should we use `autofillBgColor` to be more consistent? Idk

Comment (hasparus) [reply packages/components/src/Input.js:34]: Actually I don't think we support `bgColor` anywhere. It's either `bg` or `backgroundColor` or `background`.

**Comment by hasparus** [comment]
@flo-sch you might be interested in this PR (BTW, is there a way to contact you outside of GitHub?)

**Comment by flo-sch** [comment]
Nice feature :)

The interwebs provide many ways, are you thinking of a specific one?

**Comment by hasparus** [comment]
> The interwebs provide many ways, are you thinking of a specific one?

Okay, I've sent you an e-mail through the contact form :)

**Comment by hasparus** [comment]
:rocket: PR was released in [`v0.10.0`](https://github.com/system-ui/theme-ui/releases/tag/v0.10.0) :rocket:

**Comment by flo-sch** [comment]
Hello @hasparus, small related question before I open another issue / PR, seems like the `autofillBackgroundColor` is missing in the Input types maybe?

> Property 'autofillBackgroundColor' does not exist on type 'IntrinsicAttributes & Pick<InputProps, "name" | "form" | "p" | "slot" | "style" | "title" | "pattern" | "className" | "key" | "autoComplete" | ... 311 more ... | "step"> & RefAttributes<...> & SxProp'.ts

Also a small question/suggestion: what about defining a specific key under theme.colors so that we could directly set it as part of the theme (or a variant), instead of having to set it as a prop (on each `<Input />`)?
I was thinking about changing the prop default value:

Not sure if there is an extra smart way to achieve that, but maybe something like (potentially breaking change?):

```tsx
sx={{
  '--theme-ui-input-autofill-bg': (theme) => get(
    theme.colors,
    // first try to apply theme.colors.autofillBackgroundColor if it is defined
    'autofillBackgroundColor',
    // fallback to the prop value otherwise
    get(theme.colors, autofillBackgroundColor, null)
  ),
  ...sx,
}}
```

**Comment by hasparus** [comment]
> Hello @hasparus, small related question before I open another issue / PR, seems like the autofillBackgroundColor is missing in the Input types maybe?

Oh, that's definitely a bug. Do you wanna add it?

> Not sure if there is an extra smart way to achieve that, but maybe something like (potentially breaking change?):

To be fair, I'm not sure about it. If we're editing the theme already we might pass this color to a variant?

**Comment by flo-sch** [comment]
> Oh, that's definitely a bug. Do you wanna add it?

Yep, I can take that :)

> To be fair, I'm not sure about it. If we're editing the theme already we might pass this color to a variant?

I have tried to add it as a variant but somehow it always gets overridden by the prop (even when not set --> fallback to `background`), I think because the `<Input />` `__css` (with default style, including that autofill rule) is placed after `sx={{ ... }}` and is evaluated afterwards (as "more specific") by the browser? 🤔 

```ts
// CustomTheme.ts
{
  forms: {
    someVariant: {
      // ...
      '--theme-ui-input-autofill-bg': (theme) => get(theme.colors || {}, 'highlight', null) as string,
    }
  }
}
```

--> The default Input style overrides it anyway

theme-ui-input-autofill-bg

I could make it work by re-adding the same rules as part of the variant, bit cumbersome but it works:

```ts
// CustomTheme.ts
{
  forms: {
    someVariant: {
      // ...
      ':autofill, :autofill:hover, :autofill:focus': {
        boxShadow: (theme) =>
          `inset 0 0 0 1000px ${get(theme.colors || {}, 'highlight', 'lightgrey') as string}`
      },
      ':-webkit-autofill, :-webkit-autofill:hover, :-webkit-autofill:focus': {
        boxShadow: (theme) =>
          `inset 0 0 0 1000px ${get(theme.colors || {}, 'muted', 'lightgrey') as string}`
      }
    }
  }
}
```

(Is there any better way to type `get()`? 🤔 )

**Comment by xyzzyxyzzy0077** [comment]
Thanks for the great work! Is there any update on fixing `autofillBackgroundColor` not in the input types issue?

**Comment by lachlanjc** [comment]
Not that I'm aware of, @xyzzyxyzzy0077! @flo-sch did you end up looking into that?

**Comment by flo-sch** [comment]
Ah, I forgot about that sorry, but I should be able to very soon!

**Pull request merged**